### PR TITLE
Don't default NAT Gateway for existing node subnet

### LINKED ISSUE #3664: Egress on existing clusters using node load balancer breaks after v1.9.x upgrade
/kind bug

**What steps did you take and what happened:**

If a cluster is not using NAT gateways for the nodes after upgrading CAPZ to `v1.9.2` or `v1.9.3` new VMs no longer have egress to the Internet and never join the cluster.

1. Deploy an older version of CAPZ
```
clusterctl init --infrastructure azure:v1.8.5
```
2. Generate a cluster spec:
```
clusterctl generate cluster capi-quickstart \
  --kubernetes-version v1.27.2 \
  --control-plane-machine-count=1 \
  --worker-machine-count=2 \
  > capi-quickstart.yaml
```
3. Remove the `natGateway` config for the worker nodepool in `AzureCluster`. Should look something like:
```
  networkSpec:
    subnets:
    - name: control-plane-subnet
      role: control-plane
    - name: node-subnet
      natGateway:
        name: node-natgateway
      role: node
    vnet:
      name: capi-quickstart-vnet
```
4. Create the cluster `kubectl apply -f capi-quickstart.yaml` and follow the Quick Start to deploy [CPI](https://cluster-api.sigs.k8s.io/user/quick-start.html#install-a-cloud-provider) and [CNI](https://cluster-api.sigs.k8s.io/user/quick-start.html#deploy-a-cni-solution).
5. Upgrade CAPZ to `v1.9.2` or `v1.9.3` 
```
clusterctl upgrade apply --infrastructure azure:v1.9.3
```
6. Update the AzureCluster object in any way.
```
kubectl annotate azurecluster capi-quickstart updated=dkoskin
```
This results in the a new NAT Gateway getting created, but it never gets attached to the existing subnet.
```
      - cidrBlocks:
        - 10.1.0.0/16
        id: /subscriptions/<subscription>/resourceGroups/capi-quickstart/providers/Microsoft.Network/virtualNetworks/capi-quickstart-vnet/subnets/node-subnet
        name: node-subnet
        natGateway:
          id: /subscriptions/<subscription>/resourceGroups/capi-quickstart/providers/Microsoft.Network/natGateways/capi-quickstart-node-natgw-1
          ip:
            name: pip-capi-quickstart-node-natgw-1
          name: capi-quickstart-node-natgw-1
```

image
7. Scale up the node MachineDeployment and observe the new Node does not come up
```
kubectl scale md capi-quickstart-md-0 --replicas=3
```
Cloud init logs:
```
[  360.442475] cloud-init[1278]: [2023-06-24 05:29:11] Cloud-init v. 23.1.2-0ubuntu0~22.04.1 running 'modules:final' at Sat, 24 Jun 2023 05:29:11 +0000. Up 41.72 seconds.
[  360.449713] cloud-init[1278]: [2023-06-24 05:29:17] [preflight] Running pre-flight checks
[  360.453556] cloud-init[1278]: [2023-06-24 05:34:29] error execution phase preflight: couldn't validate the identity of the API Server: Get "https://capi-quickstart-a2a9441a.westus.cloudapp.azure.com:6443/api/v1/namespaces/kube-public/configmaps/cluster-info?timeout=10s": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
[  360.459773] cloud-init[1278]: [2023-06-24 05:34:29] To see the stack trace of this error execute with --v=5 or higher
[  360.459849] cloud-init[1278]: [2023-06-24 05:34:29] 2023-06-24 05:34:29,555 - cc_scripts_user.py[WARNING]: Failed to run module scripts-user (scripts in /var/lib/cloud/instance/scripts)
[  360.459910] cloud-init[1278]: [2023-06-24 05:34:29] 2023-06-24 05:34:29,555 - util.py[WARNING]: Running module scripts-user (<module 'cloudinit.config.cc_scripts_user' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_scripts_user.py'>) failed
[  360.459969] cloud-init[1278]: [2023-06-24 05:34:29] Cloud-init v. 23.1.2-0ubuntu0~22.04.1 finished at Sat, 24 Jun 2023 05:34:29 +0000. Datasource DataSourceAzure [seed=/dev/sr0].  Up 360.37 seconds
```

**What did you expect to happen:**
IMO existing clusters not already using a NAT Gateway for node egress should not be modified and continue not to use it, unless a user explicitly creates a new subnet.

**Anything else you would like to add:**
I see this change was made in `v1.9.0` [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3365), but I didn't observe this behavior until trying with `v1.9.2` with [this PR](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3554).

**Environment:**

- cluster-api-provider-azure version: 
`v1.9.2`
- Kubernetes version: (use `kubectl version`): 
Tried with: `v1.27.2` and `v1.26.5`
- OS (e.g. from `/etc/os-release`): 
Default Ubuntu 22.04 image

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**: If a cluster is not using NAT gateways for the nodes after upgrading CAPZ to v1.9.2 or v1.9.3 new VMs no longer have egress to the Internet and never join the cluster.

This PR:

1. Changes the defaulting to not update existing subnets
This is a bit of a hack because it relies on the ID field to determine that the defaulting is happening on an existing subnet vs a new subnet. Ideally, NAT Gateway would be a pointer so we could differentiate between disable NAT Gateway and new cluster, and we could have simply switch the default from empty string to default name, but that is not the case :(

2. We update existing subnets to "attach" the NAT Gateway if it gets added by the user later, so users can migrate to NAT Gateways for outbound connection (since it's the recommended way to do outbound connection in Azure).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3664 

**Special notes for your reviewer**:

- [x] cherry-pick candidate

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Don't default NAT Gateway for existing node subnet
```

### CODE REVIEW

**Comment by dkoshkin** [comment]
Thanks @CecileRobertMichon the changes look good, going to test on a live cluster too.

**Comment by dkoshkin** [comment]
I followed the same repro steps, then:

```
$ clusterctl upgrade apply --infrastructure azure:v1.9.3
$ kubectl set image deployment -n capz-system capz-controller-manager manager=dkoshkin/cluster-api-azure-controller-amd64:3700-1
```
And on the cluster:
```
$ kubectl annotate azurecluster capi-quickstart updated=dkoskin
$ kubectl scale md capi-quickstart-md-0 --replicas=3
```

As expected, `natGateway` was NOT added to the `AzureCluster`:
```
      - cidrBlocks:
        - 10.1.0.0/16
        id: /subscriptions/<>/resourceGroups/capi-quickstart/providers/Microsoft.Network/virtualNetworks/capi-quickstart-vnet/subnets/node-subnet
        name: node-subnet
        natGateway:
          ip:
            name: ""
          name: ""
        role: node
```
The new Node got created:
```
$ kubectl get nodes
NAME                                  STATUS   ROLES           AGE     VERSION
capi-quickstart-control-plane-rhtzx   Ready    control-plane   11m     v1.27.2
capi-quickstart-md-0-54f5s            Ready    <none>          81s     v1.27.2
capi-quickstart-md-0-kr4p2            Ready    <none>          9m24s   v1.27.2
capi-quickstart-md-0-lcql9            Ready    <none>          9m21s   v1.27.2
````

----
However, I think this introduced an unintended side-effect, since we have https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/e60f052c7a6d96cd28e130aacbf194327412b228 and the published templates don't have `natGateway` set its no longer defaulting to setting it. Should we bring it back in the templates?

**Comment by CecileRobertMichon** [comment]
> However, I think this introduced an unintended side-effect, since we have https://github.com/kubernetes-sigs/cluster-api-provider-azure/commit/e60f052c7a6d96cd28e130aacbf194327412b228 and the published templates don't have natGateway set its no longer defaulting to setting it.

@dkoshkin the defaulting should still happen for new subnets since the `ID` field will be empty at first, is that different from what you are seeing?

**Comment by dkoshkin** [comment]
Oops you're right @CecileRobertMichon, I reran it again:
```
$ git checkout release-1.9
$ git cherry-pick 1a572e8f3fc69a9cdf4f791c4a88672324f63c5b

$ make docker-build
$ docker images
REPOSITORY                                                                  TAG                                      IMAGE ID       CREATED          SIZE
gcr.io/eng-ksphere-platform-e2e/cluster-api-azure-controller-amd64          dev                                      22ef31cbb923   8 seconds ago    85.9MB

$ docker tag gcr.io/eng-ksphere-platform-e2e/cluster-api-azure-controller-amd64:dev dkoshkin/cluster-api-azure-controller-amd64:3700-1.9
$ docker push dkoshkin/cluster-api-azure-controller-amd64:3700-1.9
The push refers to repository [docker.io/dkoshkin/cluster-api-azure-controller-amd64]
54641c0c05ca: Pushed
4cb10dd2545b: Layer already exists
d2d7ec0f6756: Layer already exists
1a73b54f556b: Layer already exists
e624a5370eca: Layer already exists
d52f02c6501c: Layer already exists
ff5700ec5418: Layer already exists
7bea6b893187: Layer already exists
6fbdf253bbc2: Layer already exists
e023e0e48e6e: Layer already exists
3700-1.9: digest: sha256:f85d63e199266fdfd39308db513a7a3e0d326f1c094dbd7d5235ddbb49c0caf8 size: 2402
````

Create a new bootstrap:
```
$ kind create cluster
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.25.3) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂

$ clusterctl init --infrastructure azure
Fetching providers
Installing cert-manager Version="v1.11.1"
Waiting for cert-manager to be available...
Installing Provider="cluster-api" Version="v1.4.4" TargetNamespace="capi-system"
Installing Provider="bootstrap-kubeadm" Version="v1.4.4" TargetNamespace="capi-kubeadm-bootstrap-system"
Installing Provider="control-plane-kubeadm" Version="v1.4.4" TargetNamespace="capi-kubeadm-control-plane-system"
Installing Provider="infrastructure-azure" Version="v1.9.4" TargetNamespace="capz-system"

Your management cluster has been initialized successfully!

You can now create your first workload cluster by running the following:

  clusterctl generate cluster [name] --kubernetes-version [version] | kubectl apply -f -
```

Updated the controller to use the dev image:
```
$ kubectl set image deployment -n capz-system capz-controller-manager manager=dkoshkin/cluster-api-azure-controller-amd64:3700-1.9
deployment.apps/capz-controller-manager image updated
```

Generate cluster specs:
```
clusterctl generate cluster capi-quickstart-defaults \
  --kubernetes-version v1.27.2 \
  --control-plane-machine-count=1 \
  --worker-machine-count=2 \
  > capi-quickstart-defaults.yaml
```

The generated spec file does not have a value for `natGateway`:
```
$ cat capi-quickstart-defaults.yaml| grep -A 7 subnets
    subnets:
    - name: control-plane-subnet
      role: control-plane
    - name: node-subnet
      role: node
    vnet:
      name: capi-quickstart-defaults-vnet
  resourceGroup: capi-quickstart-defaults
```

But it gets set after it is created:
```
➜  dkkonvoy-capz-nat k get azurecluster -o yaml | grep subnets
        {"apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","kind":"AzureCluster","metadata":{"annotations":{},"name":"capi-quickstart-defaults","namespace":"default"},"spec":{"identityRef":{"apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","kind":"AzureClusterIdentity","name":"cluster-identity"},"location":"westus","networkSpec":{"subnets":[{"name":"control-plane-subnet","role":"control-plane"},{"name":"node-subnet","role":"node"}],"vnet":{"name":"capi-quickstart-defaults-vnet"}},"resourceGroup":"capi-quickstart-defaults","subscriptionID":"b1ed8f2f-ceda-41d5-a0a8-95960b5340c2"}}
      subnets:
        id: /subscriptions/b1ed8f2f-ceda-41d5-a0a8-95960b5340c2/resourceGroups/capi-quickstart-defaults/providers/Microsoft.Network/virtualNetworks/capi-quickstart-defaults-vnet/subnets/control-plane-subnet
        id: /subscriptions/b1ed8f2f-ceda-41d5-a0a8-95960b5340c2/resourceGroups/capi-quickstart-defaults/providers/Microsoft.Network/virtualNetworks/capi-quickstart-defaults-vnet/subnets/node-subnet
➜  dkkonvoy-capz-nat k get azurecluster -o yaml
apiVersion: v1
items:
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: AzureCluster
  metadata:
    name: capi-quickstart-defaults
    namespace: default
  spec:
    networkSpec:
      subnets:
      - cidrBlocks:
        - 10.0.0.0/16
        id: /subscriptions/<>/resourceGroups/capi-quickstart-defaults/providers/Microsoft.Network/virtualNetworks/capi-quickstart-defaults-vnet/subnets/control-plane-subnet
        name: control-plane-subnet
        natGateway:
          ip:
            name: ""
          name: ""
        role: control-plane
        routeTable:
          name: ""
        securityGroup:
          name: capi-quickstart-defaults-controlplane-nsg
          securityRules:
          - description: Allow SSH
            destination: '*'
            destinationPorts: "22"
            direction: Inbound
            name: allow_ssh
            priority: 2200
            protocol: Tcp
            source: '*'
            sourcePorts: '*'
          - description: Allow K8s API Server
            destination: '*'
            destinationPorts: "6443"
            direction: Inbound
            name: allow_apiserver
            priority: 2201
            protocol: Tcp
            source: '*'
            sourcePorts: '*'
      - cidrBlocks:
        - 10.1.0.0/16
        id: /subscriptions/<>/resourceGroups/capi-quickstart-defaults/providers/Microsoft.Network/virtualNetworks/capi-quickstart-defaults-vnet/subnets/node-subnet
        name: node-subnet
        natGateway:
          id: /subscriptions<>/resourceGroups/capi-quickstart-defaults/providers/Microsoft.Network/natGateways/capi-quickstart-defaults-node-natgw-1
          ip:
            name: pip-capi-quickstart-defaults-node-natgw-1
          name: capi-quickstart-defaults-node-natgw-1
        role: node
        routeTable:
          name: capi-quickstart-defaults-node-routetable
        securityGroup:
          name: capi-quickstart-defaults-node-nsg
```

**Review by dkoshkin — APPROVED** [review]
Thanks for fixing this @CecileRobertMichon!

**Comment by nawazkh** [comment]
looks good to me!
/lgtm

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Comment by CecileRobertMichon** [comment]
/retest

unrelated VMSS drain flake

**Comment by CecileRobertMichon** [comment]
> Timed out waiting for 2 ready replicas for MachinePool capz-e2e-2kja6t/capz-e2e-2kja6t-vmss-mp-win 

:(

/retest

File: api/v1beta1/azurecluster_default.go [hunk 1]
```diff
@@ -125,9 +125,10 @@ func (c *AzureCluster) setSubnetDefaults() {
 			subnet.RouteTable.Name = generateNodeRouteTableName(c.ObjectMeta.Name)
 		}
 
-		if !subnet.IsIPv6Enabled() {
-			// NAT gateway supports the use of IPv4 public IP addresses for outbound connectivity.
-			// So default use the NAT gateway for outbound traffic in IPv4 cluster instead of loadbalancer.
+		// NAT gateway only supports the use of IPv4 public IP addresses for outbound connectivity.
+		// So default use the NAT gateway for outbound traffic in IPv4 cluster instead of loadbalancer.
+		// We assume that if the ID is set, the subnet already exists so we shouldn't add a NAT gateway.
```
Comment (Jont828) [inline api/v1beta1/azurecluster_default.go:130]: Maybe we can rephrase this to say we want to add a NAT gateway only if the ID is not set. I think it would make more sense later on when people aren't looking at the context of this PR, and it would be a bit more direct in explaining what the `if` below does.

File: azure/services/subnets/spec.go [hunk 2]
```diff
@@ -135,3 +115,35 @@ func (s *SubnetSpec) Parameters(ctx context.Context, existing interface{}) (para
...
+				existingServiceEndpoints = append(existingServiceEndpoints, network.ServiceEndpointPropertiesFormat{Service: se.Service, Locations: se.Locations})
+			}
+		}
+		newServiceEndpoints := make([]network.ServiceEndpointPropertiesFormat, len(s.ServiceEndpoints))
+		for _, se := range s.ServiceEndpoints {
+			se := se
+			newServiceEndpoints = append(newServiceEndpoints, network.ServiceEndpointPropertiesFormat{Service: pointer.String(se.Service), Locations: &se.Locations})
+		}
+
+		diff := cmp.Diff(newServiceEndpoints, existingServiceEndpoints)
```
Comment (Jont828) [inline azure/services/subnets/spec.go:145]: Would it be worth it to try to log the diff somewhere? I think it would help with debugging/maintaining the unit test for this function since it'd be hard to tell what went wrong if we're just surfacing a `true` or `false` result.

**Review by Jont828 — COMMENTED** [review]
Looks good! I like that we're breaking some of the logic into the `shouldUpdate()` function and adding unit tests.

**Comment by CecileRobertMichon** [comment]
@Jont828 your review came in after the PR had already merged. Are any of the comments blocking? If not I'm happy to address them in a follow up.

**Comment by Jont828** [comment]
@CecileRobertMichon None of my review comments are blocking, just some things I noticed. The thing with `diff := cmp.Diff(newServiceEndpoints, existingServiceEndpoints)` seems to show up in other spots, I believe. I figured we could think about if it'd be worth it to refactor it since we're checking if `diff != ""` and not logging it anywhere.
